File: rxbinding/src/main/java/com/jakewharton/rxbinding2/view/ViewAttachEventObservable.java
package com.jakewharton.rxbinding2.view;

import android.view.View;
import android.view.View.OnAttachStateChangeListener;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class ViewAttachEventObservable extends Observable<ViewAttachEvent> {
  private final View view;

  ViewAttachEventObservable(View view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super ViewAttachEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.addOnAttachStateChangeListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnAttachStateChangeListener {
    private final View view;
    private final Observer<? super ViewAttachEvent> observer;

    Listener(View view, Observer<? super ViewAttachEvent> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onViewAttachedToWindow(View v) {
      if (!isDisposed()) {
        observer.onNext(ViewAttachAttachedEvent.create(view));
      }
    }

    @Override public void onViewDetachedFromWindow(View v) {
      if (!isDisposed()) {
        observer.onNext(ViewAttachDetachedEvent.create(view));
      }
    }

    @Override protected void onDispose() {
      view.removeOnAttachStateChangeListener(this);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding3/view/ViewAttachEventObservable.kt
@file:JvmName("RxView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.view

import android.view.View
import android.view.View.OnAttachStateChangeListener
import androidx.annotation.CheckResult
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of attach and detach events on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun View.attachEvents(): Observable<ViewAttachEvent> {
  return ViewAttachEventObservable(this)
}

private class ViewAttachEventObservable(
  private val view: View
) : Observable<ViewAttachEvent>() {

  override fun subscribeActual(observer: Observer<in ViewAttachEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.addOnAttachStateChangeListener(listener)
  }

  private class Listener(
    private val view: View,
    private val observer: Observer<in ViewAttachEvent>
  ) : MainThreadDisposable(), OnAttachStateChangeListener {

    override fun onViewAttachedToWindow(v: View) {
      if (!isDisposed) {
        observer.onNext(ViewAttachAttachedEvent(view))
      }
    }

    override fun onViewDetachedFromWindow(v: View) {
      if (!isDisposed) {
        observer.onNext(ViewAttachDetachedEvent(view))
      }
    }

    override fun onDispose() {
      view.removeOnAttachStateChangeListener(this)
    }
  }
}
